### PR TITLE
Mongo to V3

### PR DESCRIPTION
**Pull request by MXPOL**
### Apps: 
- [x] remove comment from [factory](https://github.com/wix/velo-external-db/blob/managed-adapter/apps/velo-external-db/src/storage/factory.ts) 
- [x] remove comment from [e2e.db.setup](https://github.com/wix/velo-external-db/blob/managed-adapter/apps/velo-external-db/test/env/env.db.setup.js)
- [x] remove comment from [e2e.db.teardown](https://github.com/wix/velo-external-db/blob/managed-adapter/apps/velo-external-db/test/env/env.db.teardown.js)
- [x] in [e2e_resources](https://github.com/wix/velo-external-db/blob/managed-adapter/apps/velo-external-db/test/resources/e2e_resources.ts) replace `createApp` with `createAppWithWixDataBaseUrl`

### Current implementation changes: 
  
  Schema:

  - [x] create [capabilities file](https://github.com/wix/velo-external-db/blob/managed-adapter/libs/external-db-mysql/src/mysql_capabilities.ts)
  - [x] change [schemas list](https://github.com/wix/velo-external-db/blob/managed-adapter/libs/external-db-mysql/src/mysql_schema_provider.ts#L28-L33) format.
  - [x] change [describe collection](https://github.com/wix/velo-external-db/blob/managed-adapter/libs/external-db-mysql/src/mysql_schema_provider.ts#L79-L93) format.
  - [x] implement [change column type](https://github.com/wix/velo-external-db/blob/managed-adapter/libs/external-db-mysql/src/mysql_schema_provider.ts#L67-L71) method.
  - [x] change [default precision](https://github.com/wix/velo-external-db/blob/managed-adapter/libs/external-db-mysql/src/sql_schema_translator.ts#L128) to be (15,2) - if relevant.
  
Data: 

- [x] support [upsert](https://github.com/wix/velo-external-db/blob/managed-adapter/libs/external-db-mysql/src/mysql_data_provider.ts#L41-L50).
- [x] add sort, skip, limit in [aggregate](https://github.com/wix/velo-external-db/blob/managed-adapter/libs/external-db-mysql/src/mysql_data_provider.ts#L77-L87) function.
- [x] support querying by nested fields (+sql_filter_transformer tests).

Error handling: 

- [x] return `collectionName` with each error code ([sql_exception_translator](https://github.com/wix/velo-external-db/blob/managed-adapter/libs/external-db-mysql/src/sql_exception_translator.ts)).
- [x] on field [doesn't exist](https://github.com/wix/velo-external-db/blob/managed-adapter/libs/external-db-mysql/src/sql_exception_translator.ts#L21), return the field name.
- [x] on field [already exist](https://github.com/wix/velo-external-db/blob/managed-adapter/libs/external-db-mysql/src/sql_exception_translator.ts#L23), return the field name.
- [x] on item [already exists](https://github.com/wix/velo-external-db/blob/managed-adapter/libs/external-db-mysql/src/sql_exception_translator.ts#L35), return the duplicated _id

e2e: 
- [x] export capabilities from [implementation_resources](https://github.com/wix/velo-external-db/blob/managed-adapter/libs/external-db-mysql/tests/e2e-testkit/mysql_resources.ts#L6)

general: 

- [x] in [workspace](https://github.com/wix/velo-external-db/blob/managed-adapter/workspace.json#L5), add the lib
- [x] in [package.json - test script](https://github.com/wix/velo-external-db/blob/managed-adapter/package.json#L10), test the current implementation too.
- [x] in [main.yml](https://github.com/wix/velo-external-db/blob/managed-adapter/.github/workflows/main.yml#L40-L42), add the current implementation.

### CODE REVIEW

File: libs/external-db-mongo/src/mongo_schema_provider.ts [hunk 1]
```diff
@@ -120,11 +155,11 @@ export default class SchemaProvider implements ISchemaProvider {
         }
     }
 
-    async collectionDataFor(collectionName: string): Promise<any> { //fixme: any
+    async collectionDataFor(collectionName: string) { //fixme: any
```
Comment (Idokah) [inline libs/external-db-mongo/src/mongo_schema_provider.ts:158]: can delete comment

File: libs/external-db-mongo/src/mongo_utils.spec.ts [hunk 2]
```diff
@@ -48,4 +48,19 @@ describe('Mongo Utils', () => {
...
+        test('insertExpressionFor with upsert set to true will return update expression', () => {
+            expect(insertExpressionFor([{ _id: 'itemId' }], true)[0]).toEqual({
+                                                                                updateOne: { 
+                                                                                            filter: { _id: 'itemId' },
+                                                                                            update: { $set: { _id: 'itemId' } },
+                                                                                            upsert: true
+                                                                                        } 
+                                                                            })
+        })
+    })
```
Comment (Idokah) [inline libs/external-db-mongo/src/mongo_utils.spec.ts:65]: 👏

File: libs/external-db-mongo/src/sql_filter_transformer.ts [hunk 3]
```diff
@@ -141,6 +141,15 @@ export default class FilterParser {
...
 
+    orderAggregationBy(sort: Sort[]) {
+        return {
+            $sort: sort.reduce((acc, s) => {
+                const direction = s.direction === 'asc'? 1 : -1 
+                return { ...acc, [s.fieldName]: direction }
+            }, {})
+        }
+    }
+
```
Comment (Idokah) [inline libs/external-db-mongo/src/sql_filter_transformer.ts:152]: add test for that in `sql_filter_transformer.spec.ts`

File: apps/velo-external-db/test/e2e/app_data.e2e.spec.ts [hunk 4]
```diff
@@ -80,7 +86,7 @@ describe(`Velo External DB Data REST API: ${currentDbImplementationName()}`,  ()
         )
     })
 
-    test('insert api should fail if item already exists', async() => {
+    testIfSupportedOperationsNotIncludes(supportedOperations, [continueInsertingOnError])('insert api should fail if item already exists', async() => {
```
Comment (Idokah) [inline apps/velo-external-db/test/e2e/app_data.e2e.spec.ts:89]: REALLY confusing, create another operation that every implementation supports but mongo- 
`AtomicBulkInsert` and IMO remove this new `testIfSupportedOperationsNotIncludes`

File: package.json [hunk 5]
```diff
@@ -7,7 +7,7 @@
     "build:full-image": "nx run velo-external-db:build-image ",
     "lint": "eslint --cache ./",
     "lint:fix": "eslint --cache --fix ./",
-    "test": "npm run test:core; npm run test:mysql; npm run test:postgres; npm run test:mssql",
+    "test": "npm run test:core; npm run test:mysql; npm run test:postgres; npm run test:mssql; npm run test:mongo; npm run test:mongo4;",
```
Comment (Idokah) [inline package.json:10]: only mongo here (look on master)

File: libs/external-db-mongo/src/mongo_data_provider.ts [hunk 6]
```diff
@@ -34,22 +34,22 @@ export default class DataProvider implements IDataProvider {
...
         validateTable(collectionName)
-        const result = await this.client.db()
-                                        .collection(collectionName)
-                                        //@ts-ignore - Type 'string' is not assignable to type 'ObjectId', objectId Can be a 24 character hex string, 12 byte binary Buffer, or a number. and we cant assume that on the _id input
-                                        .insertMany(items)
-                                        .catch(translateErrorCodes)
-        return result.insertedCount
+        const { insertedCount, upsertedCount } = await this.client.db()
+                                                                  .collection(collectionName) 
+                                                                  .bulkWrite(insertExpressionFor(items, upsert), { ordered: false })
```
[the remark this answers is not in the document]
Comment (MXPOL) [reply libs/external-db-mongo/src/mongo_data_provider.ts]: `{ ordered: false }` means that the client will continue inserting the remaining items even if one fails
 This is exactly what we want...

**Review by Idokah — APPROVED** [review]

**Pull request merged**